telegram/ext/_messagereactionhandler.py

```diff
@@ -0,0 +1,203 @@
+        super().__init__(callback, block=block)
+        self.message_reaction_types = message_reaction_types
+
+        self._chat_ids = self._parse_chat_id(chat_id)
+        self._chat_usernames = self._parse_username(chat_username)
+        if (
+            (user_id or user_username)
+            and message_reaction_types == self.MESSAGE_REACTION
+            or self.MESSAGE_REACTION_COUNT_UPDATED
+        ):
+            raise ValueError(
+                "You can not filter for users and include anonymous reactions. Set "
+                "`message_reaction_types` to MESSAGE_REACTION_COUNT_UPDATED."
+            )
+        self._user_ids = self._parse_chat_id(user_id)
+        self._user_usernames = self._parse_username(user_username)
+
+    @staticmethod
+    def _parse_chat_id(chat_id: Optional[SCT[int]]) -> FrozenSet[int]:
+        if chat_id is None:
+            return frozenset()
+        if isinstance(chat_id, int):
+            return frozenset({chat_id})
+        return frozenset(chat_id)
+
+    @staticmethod
+    def _parse_username(username: Optional[SCT[str]]) -> FrozenSet[str]:
+        if username is None:
+            return frozenset()
+        if isinstance(username, str):
+            return frozenset({username[1:] if username.startswith("@") else username})
+        return frozenset({usr[1:] if usr.startswith("@") else usr for usr in username})
+
+    # pylint: disable=too-many-return-statements
+    def check_update(self, update: object) -> bool:
+        """Determines whether an update should be passed to this handler's :attr:`callback`.
+
+        Args:
+            update (:class:`telegram.Update` | :obj:`object`): Incoming update.
+
+        Returns:
+            :obj:`bool`
+
+        """
+        if isinstance(update, Update):
+            if not (update.message_reaction or update.message_reaction_count):
+                return False
+            if (
+                self.message_reaction_types == self.MESSAGE_REACTION_UPDATED
+                and update.message_reaction_count
+            ):
+                return False
+            if (
+                self.message_reaction_types == self.MESSAGE_REACTION_COUNT_UPDATED
+                and update.message_reaction
+            ):
+                return False
+            if (
+                not self._chat_ids
+                and not self._chat_usernames
+                and not self._user_ids
+                and not self._user_usernames
+            ):
+                return True
+            # this filter must chain False, because it filters both chat and users, which wouldn't
+            # work if we return True
+            # TODO this breaks mypy and I need help with the best solution to not make it break
+            if self._chat_ids and update.effective_chat.id not in self._chat_ids:
+                return False
+            if self._chat_usernames and update.effective_chat.username not in self._chat_usernames:
+                return False
```

Comment:
I don't quite understand . from the above docstring i would have expected the following logic:

1. determine user_id, user_username, chat_id, chat_username
2. for all 4: if entity in self._corresponding_frozenset: return True
3. return False

Reply:
@Poolitzer I adapted this quite a bit now. Please double check if the new logic does what you wanted it to do.